include/triton/Tools/LinearLayout.h

```diff
@@ -0,0 +1,433 @@
+//           |bN1, bN2, ..., bNM|.
+//
+// Usually when we do linear algebra, the field 𝔽 from which `ai` and `bij` are
+// drawn is the real or complex numbers.  But in linear layouts, we let	𝔽 be a
+// different field: GF(2).
+//
+// GF(2) is the two-element field of bits.  To define a field, I need to give
+// you the set of elements and also addition and multiplication operations.  For
+// GF(2) the elements are simply {0,1}.  We define addition as xor, and
+// multiplication as binary `and`.
+//
+// Here's an example of a 4x4 matrix-vector multiply where the elements are in
+// GF(2).  I'm using ⊕ to represent GF(2)'s addition operation (i.e xor) and ×
+// to represent multiplication (i.e. binary `and`).
+//
+//    | 1 0 0 0 | | 0 |     | 1 |         | 0 |         | 0 |         | 0 |
+//    | 0 1 1 0 | | 1 |  =  | 0 | × 0  ⊕  | 1 | × 1  ⊕  | 1 | × 1  ⊕  | 0 | × 0
+//    | 0 0 1 1 | | 1 |     | 0 |         | 0 |         | 1 |         | 1 |
+//    | 0 0 1 1 | | 0 |     | 0 |         | 0 |         | 1 |         | 1 |
+//
+//                                        | 0 |         | 0 |
+//                       =                | 1 |    ⊕    | 1 |
+//                                        | 0 |         | 1 |
+//                                        | 0 |         | 1 |
+//
+//                          | 0 |
+//                       =  | 0 |.
+//                          | 1 |
+//                          | 1 |
+//
+// This works, but it's cumbersome.  It's more compact to think of the vector
+// `a` as an M-bit unsigned integer, and each column Bi of the matrix B as an
+// N-bit integer.  Here's the same matrix-vector product written this way.
+//
+//   = | 1 2 14 12 | × 6
+//   = | 1 2 14 12 | × 0b0110
+//   = (1 × 0) ⊕ (2 × 1) ⊕ (14 × 1) ⊕ (12 × 0)
+//   = 2 ⊕ 14
+//   = 12.
+//
+// And we confirm that our answer of 12 is equal to the binary value 0b1100 we
+// got before.
+//
+// Notice that the function F(a) is fully specified by the matrix B, and that
+// the four columns of B tell us the values of F at power-of-two values for `a`,
+// namely F(1), F(2), F(4), and F(8).  In other words, we specify four values of
+// F (we call these the function's "basis vectors" or its "bases") and we can
+// then compute any other value by xor'ing together subsets of the bases.
+//
+// In the case of a 1D -> 1D layout, the implementation of an LL is
+// straightforward from the mathematical description.  If the LL is
+// higher-dimensional, we can "stack" the bit vectors to create 1D vectors.  For
+// example, if we have a 2D LL and we're given input tuple (0b0011, 0b1100), we
+// can treat this like a 1D input 0b0011'1100 and then do the regular 1D LL
+// computation.  Similarly we can "unstack" the output from 1D to ND.
+//
+// With stacking and unstacking, the linearity of our function in GF(2) directly
+// implies the "linearity rule" from earlier.  And that's all we need in order
+// to define linear layouts mathematically!
```

Comment:
To finish up the mathematical formalism, here we are defining linear functions on F_n^k (k direct sums on F_n). When I was reading through this, I was getting a bit confused because often you define k-linear functions on F_n^k, and the "linearity formula" above is actually a bilinearity formula. But of course, here you just want plain ol' linearity with the direct sum rather than the tensor product, where the basis of the resulting space is the concatenation of both bases.

Reply:
I tried to incorporate this without adding a ton more formalism (mostly so I don't have to refresh all of these definitions).  LMK if it's enough not to be confusing.

Thanks again for your feedback.